src/wormhole_mailbox_server/server.py

```diff
@@ -542,9 +543,100 @@ def _shutdown(self):
             channel._shutdown()
 
 
+def leading_zero_bits(bytestring):
+    """
+    :returns int: the number of leading zeros in the given byte-string
+    """
+    measured_bits = 0
+    for byte in bytestring:
+        bit = 1 << 7
+        while bit:
+            if byte & bit:
+                return measured_bits
+            else:
+                measured_bits += 1
+            bit = bit >> 1
+
+
+class NoPermission(object):
+    """
+    A no-op permission provider used to grant any client access (the
+    default).
+    """
+    name = "none"
+
+    def get_welcome_data(self):
+        return {}
+
+    def verify_permission(self, submit_permission):
+        return True
+
+    def is_passed(self):
+        return True
+
+
+class HashcashPermission(object):
+    """
+    A permission provider that generates a random 'resource' string
+    and checks a proof-of-work from the client.
+    """
+    name = "hashcash"
+
+    def __init__(self, bits=20):
+        self._bits = bits
+        self._passed = False
+
+    def get_welcome_data(self):
+        """
+        Generate the data to include under this method's key in the
+        `permission-required` value of the welcome message.
+
+        Should be called at most once per connection.
+        """
+        self._hashcash_resource = base64.b64encode(os.urandom(8)).decode("utf8")
+        return {
+            "bits": self._bits,
+            "resource": self._hashcash_resource,
+        }
+
+    def is_passed(self):
+        """
+        :returns bool: True if verify_permission has been called successfully
+        """
+        return self._passed
+
+    def verify_permission(self, perms):
+        """
+        :returns bool: an indication of whether the provided permissions
+            reply from a client is valid
+        """
+        stamp = perms.get("stamp", "")
+        fields = stamp.split(":")
+        if len(fields) != 7:
+            return False
+        vers, claimed_bits, date, resource, ext, rand, counter = fields
+        vers = int(vers)
+        if vers != 1:
+            return False
+        if resource != self._hashcash_resource:
+            return False
+
+        claimed_bits = int(claimed_bits)
+        if claimed_bits < self._bits:
+            return False
+
+        h = hashlib.sha1()
+        h.update(stamp.encode("utf8"))
+        measured_hash = h.digest()
+        if leading_zero_bits(measured_hash) < claimed_bits:
+            return False
+        self._passed = True
+        return True
+
+
 class Server(service.MultiService):
     def __init__(self, db, allow_list, welcome,
-                 blur_usage, usage_db=None, log_file=None):
+                 blur_usage, usage_db=None, log_file=None, permissions="none"):
```

Comment:
We should allow for more than one permission system at once. (Default value `["none"]` or `["none", "hashcash"]`)

Reply:
Currently the server only supports one at a time.
What would it mean for the server to be configured with both "none" _and_ "hashcash"?

Reply:
Normally, there's no incentive for clients to do a hashcash unless you're forced to do. But I'm inclined to using that infrastructure anyways just to make sure that code path is well tested, in the hope that it eases transition.

Reply:
That sounds like a feature request...? (i.e. "configure the hashcash bits / difficulty"). Not that I'm opposed, but also don't want to make this any more complex yet.

Clients should only send hashcash when the server asks -- and it'll only ask if `permissions="hashcash"` here. Clients that don't know about hashcash obviously can't send it, so always asking isn't going to work out well on any default server deployment (that's why the default is `"none"`).

Reply:
oh,  you actually mean a different feature: having the server send `"permissions": {"none": {}, "hashcash": {...}}` and then clients do "whatever". (You'd still have to turn _off_ the "none" one if under load etc)

Reply:
I agree that it would make more sense for when we have more than one non-trivial permission system, for example hashcash and some auth/login or multiple different PoW systems.

Reply:
I've filed https://github.com/magic-wormhole/magic-wormhole-mailbox-server/issues/30 for now to avoid making this PR more complex